### PR TITLE
Implementation of simple sorting algorithms

### PR DESCRIPTION
**Pull request by isnullxbh**

### CODE REVIEW

**Review by thequickf — APPROVED** [review]
TL;DR
Seems alright

File: include/dsa/sorting/merge_insertion_sort.hpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+template<typename Iterator,
+         typename Comparator = std::less<>>
+auto mergeInsertionSort(Iterator beg, Iterator end, const Comparator& comparator = {}) -> void
+{
+    static constexpr std::size_t threshold = 15;
+
+    if (std::distance(beg, end) <= 1) return;
+
+    const auto len = static_cast<std::size_t>(std::distance(beg, end));
+    const auto mid = static_cast<std::size_t>(std::floor(len / 2));
```
[the remark this answers is not in the document]
Comment (isnullxbh) [reply include/dsa/sorting/merge_insertion_sort.hpp:41]: I prefer a way with auto and static_cast.

File: include/dsa/sorting/merge_insertion_sort.hpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+ *
+ * @see     https://en.wikipedia.org/wiki/Merge-insertion_sort
+ * @see     M. Peczarski, The Ford–Johnson algorithm still unbeaten for less than 47 elements
+ */
+
+template<typename Iterator,
+         typename Comparator = std::less<>>
+auto mergeInsertionSort(Iterator beg, Iterator end, const Comparator& comparator = {}) -> void
+{
+    static constexpr std::size_t threshold = 15;
```
[the remark this answers is not in the document]
Comment (isnullxbh) [reply include/dsa/sorting/merge_insertion_sort.hpp:36]: Ok, but under what condition should I do insertion sort?

File: include/dsa/sorting/merge_insertion_sort.hpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+ *
+ * @tparam  Iterator    The type of the iterator.
+ * @tparam  Comparator  The type of the comparator to be used to compare keys.
+ * @param   beg         An iterator pointing to the first element of the sequence to sort.
+ * @param   end         An iterator pointing to the element following to the last element of the sequence to sort.
+ * @param   comparator  A reference to comparator.
+ *
+ * @since   0.1.0
+ *
+ * @see     https://en.wikipedia.org/wiki/Merge-insertion_sort
```
[the remark this answers is not in the document]
Comment (isnullxbh) [reply include/dsa/sorting/merge_insertion_sort.hpp]: Fixed

File: include/dsa/sorting/merge_insertion_sort.hpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+ * @see     M. Peczarski, The Ford–Johnson algorithm still unbeaten for less than 47 elements
```
[the remark this answers is not in the document]
Comment (isnullxbh) [reply include/dsa/sorting/merge_insertion_sort.hpp]: Fixed

File: include/dsa/sorting/insertion_sort.hpp [hunk 5]
```diff
@@ -9,10 +9,24 @@
 
 #include <type_traits>
 #include <iterator>
-#include <functional> // for std::greater
+#include <functional> // for std::less
 
 namespace dsa {
 
+/**
+ * @brief   Insertion sort.
```
[the remark this answers is not in the document]
Comment (isnullxbh) [reply include/dsa/sorting/insertion_sort.hpp:17]: tmp is an iterator, not key.

File: include/dsa/sorting/merge_sort.hpp [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    for (auto cur = std::next(beg, mid); cur != end; cur = std::next(cur))
+        r.push_back(std::move(*cur));
+
+    assert(l.size() + r.size() == len);
+
+    auto lcur = l.begin(), rcur = r.begin(), cur = beg;
+    for (;cur != end && lcur != l.end() && rcur != r.end(); cur = std::next(cur))
+    {
+        if (comparator(*lcur, *rcur))
+        {
```
Comment (thequickf) [inline include/dsa/sorting/merge_sort.hpp:62]: excessive curly braces?

File: libs/bsl/src/sorting.cpp [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        while (vec[j] > key)
+        {
+            vec[j + 1] = vec[j];
+            if (j-- == 0) break;
+        }
+        vec[j + 1] = key;
+    }
+}
+
+auto mergeSortNoIterators(std::vector<int>& vec, std::size_t beg, std::size_t end) -> void
```
[the remark this answers is not in the document]
Comment (isnullxbh) [reply libs/bsl/src/sorting.cpp:28]: No, then I would have to implement one more function that accepts vector and range (to call it recursively).

**Comment by thequickf** [comment]
I would suggest to add some fuzzing tests

File: include/dsa/sorting/merge_sort.hpp (hunk 6 above)
Comment (isnullxbh) [reply include/dsa/sorting/merge_sort.hpp:62]: Fixed

**Comment by isnullxbh** [comment]
@thequickf create an issue, please, I will add it in the next iteration.

File: include/dsa/sorting/merge_insertion_sort.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (thequickf) [reply include/dsa/sorting/merge_insertion_sort.hpp:36]: Sorry, It's old comment. `threshold` is ok in current version

File: include/dsa/sorting/merge_insertion_sort.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (thequickf) [reply include/dsa/sorting/merge_insertion_sort.hpp:41]: I don't like `std::floor`, doesn't do anything.

File: include/dsa/sorting/insertion_sort.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (thequickf) [reply include/dsa/sorting/insertion_sort.hpp:17]: You do store tmp value in `val`. Isn't it? `std::swap` do the same but looks better in my opinion.

[the remark this answers is not in the document]
Comment (isnullxbh) [reply include/dsa/sorting/insertion_sort.hpp:17]: I store the current key in the `val`. I do `std::move` to support non-copyable types.

**Comment by thequickf** [comment]
> @thequickf create an issue, please, I will add it in the next iteration.

Probably I'll do the PR if I will be able to find some time.

File: libs/bsl/src/sorting.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (thequickf) [reply libs/bsl/src/sorting.cpp:28]: Indeed. Implement one more implementation function and hide it in cpp file. Public function will just call it with correct beg and end.

File: include/dsa/sorting/insertion_sort.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (thequickf) [reply include/dsa/sorting/insertion_sort.hpp:17]: I'm sure that this implementation is as fast as swap, but it looks messy in my opinion.

**Pull request merged**